// js-sdsl/dist/esm/container/TreeContainer/Base/TreeBaseContainer.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import { Container } from "../../ContainerBase/index";
import { checkWithinAccessParams } from "../../../utils/checkParams";
import TreeNode from './TreeNode';
var TreeBaseContainer = /** @class */ (function (_super) {
    __extends(TreeBaseContainer, _super);
    function TreeBaseContainer(cmp) {
        if (cmp === void 0) { cmp = function (x, y) {
            if (x < y)
                return -1;
            if (x > y)
                return 1;
            return 0;
        }; }
        var _this = _super.call(this) || this;
        _this.root = undefined;
        _this.header = new TreeNode();
        _this.inOrderTraversal = function (curNode, callback) {
            if (curNode === undefined)
                return false;
            var ifReturn = _this.inOrderTraversal(curNode.left, callback);
            if (ifReturn)
                return true;
            if (callback(curNode))
                return true;
            return _this.inOrderTraversal(curNode.right, callback);
        };
        _this.cmp = cmp;
        return _this;
    }
    TreeBaseContainer.prototype._lowerBound = function (curNode, key) {
        var resNode;
        while (curNode) {
            var cmpResult = this.cmp(curNode.key, key);
            if (cmpResult < 0) {
                curNode = curNode.right;
            }
            else if (cmpResult > 0) {
                resNode = curNode;
                curNode = curNode.left;
            }
            else
                return curNode;
        }
        return resNode === undefined ? this.header : resNode;
    };
    TreeBaseContainer.prototype._upperBound = function (curNode, key) {
        var resNode;
        while (curNode) {
            var cmpResult = this.cmp(curNode.key, key);
            if (cmpResult <= 0) {
                curNode = curNode.right;
            }
            else if (cmpResult > 0) {
                resNode = curNode;
                curNode = curNode.left;
            }
        }
        return resNode === undefined ? this.header : resNode;
    };
    TreeBaseContainer.prototype._reverseLowerBound = function (curNode, key) {
        var resNode;
        while (curNode) {
            var cmpResult = this.cmp(curNode.key, key);
            if (cmpResult < 0) {
                resNode = curNode;
                curNode = curNode.right;
            }
            else if (cmpResult > 0) {
                curNode = curNode.left;
            }
            else
                return curNode;
        }
        return resNode === undefined ? this.header : resNode;
    };
    TreeBaseContainer.prototype._reverseUpperBound = function (curNode, key) {
        var resNode;
        while (curNode) {
            var cmpResult = this.cmp(curNode.key, key);
            if (cmpResult < 0) {
                resNode = curNode;
                curNode = curNode.right;
            }
            else if (cmpResult >= 0) {
                curNode = curNode.left;
            }
        }
        return resNode === undefined ? this.header : resNode;
    };
    TreeBaseContainer.prototype.eraseNodeSelfBalance = function (curNode) {
        while (true) {
            var parentNode = curNode.parent;
            if (parentNode === this.header)
                return;
            if (curNode.color === TreeNode.RED) {
                curNode.color = TreeNode.BLACK;
                return;
            }
            if (curNode === parentNode.left) {
                var brother = parentNode.right;
                if (brother.color === TreeNode.RED) {
                    brother.color = TreeNode.BLACK;
                    parentNode.color = TreeNode.RED;
                    if (parentNode === this.root) {
                        this.root = parentNode.rotateLeft();
                    }
                    else
                        parentNode.rotateLeft();
                }
                else if (brother.color === TreeNode.BLACK) {
                    if (brother.right && brother.right.color === TreeNode.RED) {
                        brother.color = parentNode.color;
                        parentNode.color = TreeNode.BLACK;
                        brother.right.color = TreeNode.BLACK;
                        if (parentNode === this.root) {
                            this.root = parentNode.rotateLeft();
                        }
                        else
                            parentNode.rotateLeft();
                        return;
                    }
                    else if (brother.left && brother.left.color === TreeNode.RED) {
                        brother.color = TreeNode.RED;
                        brother.left.color = TreeNode.BLACK;
                        brother.rotateRight();
                    }
                    else {
                        brother.color = TreeNode.RED;
                        curNode = parentNode;
                    }
                }
            }
            else {
                var brother = parentNode.left;
                if (brother.color === TreeNode.RED) {
                    brother.color = TreeNode.BLACK;
                    parentNode.color = TreeNode.RED;
                    if (parentNode === this.root) {
                        this.root = parentNode.rotateRight();
                    }
                    else
                        parentNode.rotateRight();
                }
                else {
                    if (brother.left && brother.left.color === TreeNode.RED) {
                        brother.color = parentNode.color;
                        parentNode.color = TreeNode.BLACK;
                        brother.left.color = TreeNode.BLACK;
                        if (parentNode === this.root) {
                            this.root = parentNode.rotateRight();
                        }
                        else
                            parentNode.rotateRight();
                        return;
                    }
                    else if (brother.right && brother.right.color === TreeNode.RED) {
                        brother.color = TreeNode.RED;
                        brother.right.color = TreeNode.BLACK;
                        brother.rotateLeft();
                    }
                    else {
                        brother.color = TreeNode.RED;
                        curNode = parentNode;
                    }
                }
            }
        }
    };
    TreeBaseContainer.prototype.eraseNode = function (curNode) {
        var _a, _b;
        if (this.length === 1) {
            this.clear();
            return this.header;
        }
        var swapNode = curNode;
        while (swapNode.left || swapNode.right) {
            if (swapNode.right) {
                swapNode = swapNode.right;
                while (swapNode.left)
                    swapNode = swapNode.left;
            }
            else if (swapNode.left) {
                swapNode = swapNode.left;
            }
            _a = __read([swapNode.key, curNode.key], 2), curNode.key = _a[0], swapNode.key = _a[1];
            _b = __read([swapNode.value, curNode.value], 2), curNode.value = _b[0], swapNode.value = _b[1];
            curNode = swapNode;
        }
        if (this.header.left === swapNode) {
            this.header.left = swapNode.parent;
        }
        else if (this.header.right === swapNode) {
            this.header.right = swapNode.parent;
        }
        this.eraseNodeSelfBalance(swapNode);
        swapNode.remove();
        this.length -= 1;
        this.root.color = TreeNode.BLACK;
    };
    TreeBaseContainer.prototype.insertNodeSelfBalance = function (curNode) {
        while (true) {
            var parentNode = curNode.parent;
            if (parentNode.color === TreeNode.BLACK)
                return;
            var grandParent = parentNode.parent;
            if (parentNode === grandParent.left) {
                var uncle = grandParent.right;
                if (uncle && uncle.color === TreeNode.RED) {
                    uncle.color = parentNode.color = TreeNode.BLACK;
                    if (grandParent === this.root)
                        return;
                    grandParent.color = TreeNode.RED;
                    curNode = grandParent;
                    continue;
                }
                else if (curNode === parentNode.right) {
                    curNode.color = TreeNode.BLACK;
                    if (curNode.left)
                        curNode.left.parent = parentNode;
                    if (curNode.right)
                        curNode.right.parent = grandParent;
                    parentNode.right = curNode.left;
                    grandParent.left = curNode.right;
                    curNode.left = parentNode;
                    curNode.right = grandParent;
                    if (grandParent === this.root) {
                        this.root = curNode;
                        this.header.parent = curNode;
                    }
                    else {
                        var GP = grandParent.parent;
                        if (GP.left === grandParent) {
                            GP.left = curNode;
                        }
                        else
                            GP.right = curNode;
                    }
                    curNode.parent = grandParent.parent;
                    parentNode.parent = curNode;
                    grandParent.parent = curNode;
                }
                else {
                    parentNode.color = TreeNode.BLACK;
                    if (grandParent === this.root) {
                        this.root = grandParent.rotateRight();
                    }
                    else
                        grandParent.rotateRight();
                }
                grandParent.color = TreeNode.RED;
            }
            else {
                var uncle = grandParent.left;
                if (uncle && uncle.color === TreeNode.RED) {
                    uncle.color = parentNode.color = TreeNode.BLACK;
                    if (grandParent === this.root)
                        return;
                    grandParent.color = TreeNode.RED;
                    curNode = grandParent;
                    continue;
                }
                else if (curNode === parentNode.left) {
                    curNode.color = TreeNode.BLACK;
                    if (curNode.left)
                        curNode.left.parent = grandParent;
                    if (curNode.right)
                        curNode.right.parent = parentNode;
                    grandParent.right = curNode.left;
                    parentNode.left = curNode.right;
                    curNode.left = grandParent;
                    curNode.right = parentNode;
                    if (grandParent === this.root) {
                        this.root = curNode;
                        this.header.parent = curNode;
                    }
                    else {
                        var GP = grandParent.parent;
                        if (GP.left === grandParent) {
                            GP.left = curNode;
                        }
                        else
                            GP.right = curNode;
                    }
                    curNode.parent = grandParent.parent;
                    parentNode.parent = curNode;
                    grandParent.parent = curNode;
                }
                else {
                    parentNode.color = TreeNode.BLACK;
                    if (grandParent === this.root) {
                        this.root = grandParent.rotateLeft();
                    }
                    else
                        grandParent.rotateLeft();
                }
                grandParent.color = TreeNode.RED;
            }
            return;
        }
    };
    TreeBaseContainer.prototype.findElementNode = function (curNode, key) {
        while (curNode) {
            var cmpResult = this.cmp(curNode.key, key);
            if (cmpResult < 0) {
                curNode = curNode.right;
            }
            else if (cmpResult > 0) {
                curNode = curNode.left;
            }
            else
                return curNode;
        }
        return curNode;
    };
    TreeBaseContainer.prototype.set = function (key, value) {
        if (this.root === undefined) {
            this.length += 1;
            this.root = new TreeNode(key, value);
            this.root.color = TreeNode.BLACK;
            this.root.parent = this.header;
            this.header.parent = this.root;
            this.header.left = this.root;
            this.header.right = this.root;
            return;
        }
        var curNode;
        var minNode = this.header.left;
        var compareToMin = this.cmp(minNode.key, key);
        if (compareToMin === 0) {
            minNode.value = value;
            return;
        }
        else if (compareToMin > 0) {
            minNode.left = new TreeNode(key, value);
            minNode.left.parent = minNode;
            curNode = minNode.left;
            this.header.left = curNode;
        }
        else {
            var maxNode = this.header.right;
            var compareToMax = this.cmp(maxNode.key, key);
            if (compareToMax === 0) {
                maxNode.value = value;
                return;
            }
            else if (compareToMax < 0) {
                maxNode.right = new TreeNode(key, value);
                maxNode.right.parent = maxNode;
                curNode = maxNode.right;
                this.header.right = curNode;
            }
            else {
                curNode = this.root;
                while (true) {
                    var cmpResult = this.cmp(curNode.key, key);
                    if (cmpResult > 0) {
                        if (curNode.left === undefined) {
                            curNode.left = new TreeNode(key, value);
                            curNode.left.parent = curNode;
                            curNode = curNode.left;
                            break;
                        }
                        curNode = curNode.left;
                    }
                    else if (cmpResult < 0) {
                        if (curNode.right === undefined) {
                            curNode.right = new TreeNode(key, value);
                            curNode.right.parent = curNode;
                            curNode = curNode.right;
                            break;
                        }
                        curNode = curNode.right;
                    }
                    else {
                        curNode.value = value;
                        return;
                    }
                }
            }
        }
        this.length += 1;
        this.insertNodeSelfBalance(curNode);
    };
    TreeBaseContainer.prototype.clear = function () {
        this.length = 0;
        this.root = undefined;
        this.header.parent = undefined;
        this.header.left = this.header.right = undefined;
    };
    TreeBaseContainer.prototype.eraseElementByPos = function (pos) {
        var _this = this;
        checkWithinAccessParams(pos, 0, this.length - 1);
        var index = 0;
        this.inOrderTraversal(this.root, function (curNode) {
            if (pos === index) {
                _this.eraseNode(curNode);
                return true;
            }
            index += 1;
            return false;
        });
    };
    /**
     * Removes the elements of the specified key.
     */
    TreeBaseContainer.prototype.eraseElementByKey = function (key) {
        if (!this.length)
            return;
        var curNode = this.findElementNode(this.root, key);
        if (curNode === undefined)
            return;
        this.eraseNode(curNode);
    };
    /**
     * @return An iterator point to the next iterator.
     * Removes element by iterator.
     */
    TreeBaseContainer.prototype.eraseElementByIterator = function (iter) {
        // @ts-ignore
        var node = iter.node;
        if (node === this.header) {
            throw new RangeError('Invalid iterator');
        }
        if (node.right === undefined) {
            iter = iter.next();
        }
        this.eraseNode(node);
        return iter;
    };
    /**
     * @return The height of the RB-tree.
     */
    TreeBaseContainer.prototype.getHeight = function () {
        if (!this.length)
            return 0;
        var traversal = function (curNode) {
            if (!curNode)
                return 1;
            return Math.max(traversal(curNode.left), traversal(curNode.right)) + 1;
        };
        return traversal(this.root);
    };
    return TreeBaseContainer;
}(Container));
export default TreeBaseContainer;
