// js-sdsl/dist/esm/container/HashContainer/HashSet.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
import OrderedSet from '../TreeContainer/OrderedSet';
import HashContainerBase from './Base/index';
import Vector from '../SequentialContainer/Vector';
var HashSet = /** @class */ (function (_super) {
    __extends(HashSet, _super);
    function HashSet(container, initBucketNum, hashFunc) {
        if (container === void 0) { container = []; }
        var _this = _super.call(this, initBucketNum, hashFunc) || this;
        _this.hashTable = [];
        container.forEach(function (element) { return _this.insert(element); });
        return _this;
    }
    HashSet.prototype.reAllocate = function () {
        var _this = this;
        if (this.bucketNum >= HashSet.maxBucketNum)
            return;
        var newHashTable = [];
        var originalBucketNum = this.bucketNum;
        this.bucketNum <<= 1;
        var _loop_1 = function (index) {
            var container = this_1.hashTable[index];
            if (!container || container.empty())
                return "continue";
            if (container.size() === 1) {
                var element = container.front();
                newHashTable[this_1.hashFunc(element) & (this_1.bucketNum - 1)] = new Vector([element], false);
                return "continue";
            }
            var lowList = [];
            var highList = [];
            container.forEach(function (element) {
                var hashCode = _this.hashFunc(element);
                if ((hashCode & originalBucketNum) === 0) {
                    lowList.push(element);
                }
                else
                    highList.push(element);
            });
            if (container instanceof OrderedSet) {
                if (lowList.length > HashSet.untreeifyThreshold) {
                    newHashTable[index] = new OrderedSet(lowList);
                }
                else if (lowList.length) {
                    newHashTable[index] = new Vector(lowList, false);
                }
                if (highList.length > HashSet.untreeifyThreshold) {
                    newHashTable[index + originalBucketNum] = new OrderedSet(highList);
                }
                else if (highList.length) {
                    newHashTable[index + originalBucketNum] = new Vector(highList, false);
                }
            }
            else {
                if (lowList.length >= HashSet.treeifyThreshold) {
                    newHashTable[index] = new OrderedSet(lowList);
                }
                else if (lowList.length) {
                    newHashTable[index] = new Vector(lowList, false);
                }
                if (highList.length >= HashSet.treeifyThreshold) {
                    newHashTable[index + originalBucketNum] = new OrderedSet(highList);
                }
                else if (highList.length) {
                    newHashTable[index + originalBucketNum] = new Vector(highList, false);
                }
            }
        };
        var this_1 = this;
        for (var index = 0; index < originalBucketNum; ++index) {
            _loop_1(index);
        }
        this.hashTable = newHashTable;
    };
    HashSet.prototype.clear = function () {
        this.length = 0;
        this.bucketNum = this.initBucketNum;
        this.hashTable = [];
    };
    HashSet.prototype.forEach = function (callback) {
        var index = 0;
        for (var i = 0; i < this.bucketNum; ++i) {
            if (!this.hashTable[i])
                continue;
            this.hashTable[i].forEach(function (element) { return callback(element, index++); });
        }
    };
    /**
     * Inserts element to Set.
     */
    HashSet.prototype.insert = function (element) {
        var index = this.hashFunc(element) & (this.bucketNum - 1);
        if (!this.hashTable[index]) {
            this.hashTable[index] = new Vector([element], false);
            this.length += 1;
        }
        else {
            var preSize = this.hashTable[index].size();
            if (this.hashTable[index] instanceof Vector) {
                if (!this.hashTable[index].find(element)
                    .equals(this.hashTable[index].end()))
                    return;
                this.hashTable[index].pushBack(element);
                if (this.bucketNum <= HashSet.minTreeifySize) {
                    this.length += 1;
                    this.reAllocate();
                    return;
                }
                else if (this.hashTable[index].size() >= HashSet.treeifyThreshold) {
                    this.hashTable[index] = new OrderedSet(this.hashTable[index]);
                }
            }
            else
                this.hashTable[index].insert(element);
            var curSize = this.hashTable[index].size();
            this.length += curSize - preSize;
        }
        if (this.length > this.bucketNum * HashSet.sigma) {
            this.reAllocate();
        }
    };
    /**
     * Removes the elements of the specified value.
     */
    HashSet.prototype.eraseElementByKey = function (element) {
        var index = this.hashFunc(element) & (this.bucketNum - 1);
        if (!this.hashTable[index])
            return;
        var preSize = this.hashTable[index].size();
        if (this.hashTable[index] instanceof Vector) {
            this.hashTable[index].eraseElementByValue(element);
        }
        else {
            this.hashTable[index].eraseElementByKey(element);
        }
        if (this.hashTable[index] instanceof OrderedSet) {
            if (this.hashTable[index].size() <= HashSet.untreeifyThreshold) {
                this.hashTable[index] = new Vector(this.hashTable[index]);
            }
        }
        var curSize = this.hashTable[index].size();
        this.length += curSize - preSize;
    };
    /**
     * @return If the specified element in the HashSet.
     */
    HashSet.prototype.find = function (element) {
        var index = this.hashFunc(element) & (this.bucketNum - 1);
        if (!this.hashTable[index])
            return false;
        return !this.hashTable[index].find(element)
            .equals(this.hashTable[index].end());
    };
    /**
     * Using for 'for...of' syntax like Array.
     */
    HashSet.prototype[Symbol.iterator] = function () {
        return function () {
            var index, _a, _b, element, e_1_1;
            var e_1, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        index = 0;
                        _d.label = 1;
                    case 1:
                        if (!(index < this.bucketNum)) return [3 /*break*/, 10];
                        while (index < this.bucketNum && !this.hashTable[index])
                            index += 1;
                        if (index >= this.bucketNum)
                            return [3 /*break*/, 10];
                        _d.label = 2;
                    case 2:
                        _d.trys.push([2, 7, 8, 9]);
                        _a = (e_1 = void 0, __values(this.hashTable[index])), _b = _a.next();
                        _d.label = 3;
                    case 3:
                        if (!!_b.done) return [3 /*break*/, 6];
                        element = _b.value;
                        return [4 /*yield*/, element];
                    case 4:
                        _d.sent();
                        _d.label = 5;
                    case 5:
                        _b = _a.next();
                        return [3 /*break*/, 3];
                    case 6: return [3 /*break*/, 9];
                    case 7:
                        e_1_1 = _d.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 9];
                    case 8:
                        try {
                            if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 9:
                        index += 1;
                        return [3 /*break*/, 1];
                    case 10: return [2 /*return*/];
                }
            });
        }.bind(this)();
    };
    return HashSet;
}(HashContainerBase));
export default HashSet;
